processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java

```diff
@@ -90,36 +99,66 @@ public TopNParams makeInitParams(
   public void scanAndAggregate(
       TopNParams params,
       Aggregator[][] rowSelector,
-      Map<String, Aggregator[]> aggregatesStore,
+      Map<Comparable, Aggregator[]> aggregatesStore,
       int numProcessed
   )
   {
     final Cursor cursor = params.getCursor();
     final DimensionSelector dimSelector = params.getDimSelector();
 
-    while (!cursor.isDone()) {
-      final IndexedInts dimValues = dimSelector.getRow();
-
-      for (int i = 0; i < dimValues.size(); ++i) {
-
-        final int dimIndex = dimValues.get(i);
-        Aggregator[] theAggregators = rowSelector[dimIndex];
-        if (theAggregators == null) {
-          final String key = dimSelector.lookupName(dimIndex);
-          theAggregators = aggregatesStore.get(key);
+    if (columnCapabilities.isDictionaryEncoded()) {
+      while (!cursor.isDone()) {
+        final IndexedInts dimValues = dimSelector.getRow();
+        for (int i = 0; i < dimValues.size(); ++i) {
+          final int dimIndex = dimValues.get(i);
+          Aggregator[] theAggregators = rowSelector[dimIndex];
           if (theAggregators == null) {
-            theAggregators = makeAggregators(cursor, query.getAggregatorSpecs());
-            aggregatesStore.put(key, theAggregators);
+            final String key = dimSelector.lookupName(dimIndex);
+            theAggregators = aggregatesStore.get(key);
+            if (theAggregators == null) {
+              theAggregators = makeAggregators(cursor, query.getAggregatorSpecs());
+              aggregatesStore.put(key, theAggregators);
+            }
+            rowSelector[dimIndex] = theAggregators;
           }
-          rowSelector[dimIndex] = theAggregators;
-        }
 
+          for (Aggregator aggregator : theAggregators) {
+            aggregator.aggregate();
+          }
+        }
+        cursor.advance();
+      }
+    } else {
+      while (!cursor.isDone()) {
+        // Use the aggregatesStore directly since there is no encoding
+        Comparable key;
+        switch(columnCapabilities.getType()) {
```

Comment:
can we move this logic outside the cursor loop? should help to improve performance